telefunc/node/server/shield/codegen/generateShield.ts

```diff
@@ -0,0 +1,379 @@
+}
+
+function printFailures(root: string) {
+  const failures = generatedShields.filter((s) => s.failed)
+
+  const projects = unique(failures.map((f) => f.project))
+
+  let hasTypeScriptErrors = false
+  projects.forEach((project) => {
+    const diagnostics = project.getPreEmitDiagnostics()
+    if (diagnostics.length > 0) {
+      hasTypeScriptErrors = true
+      if (project.tsConfigFilePath) {
+        console.log(`TypeScript project ${pc.bold(project.tsConfigFilePath)} errors:`)
+      } else {
+        console.log(`TypeScript errors:`)
+      }
+      console.log(project.formatDiagnosticsWithColorAndContext(diagnostics))
+    }
+  })
+
+  const docsLink = 'https://telefunc.com/shield#typescript'
+  assertWarning(
+    failures.length === 0,
+    [
+      'Failed to generate shield() for telefunction',
+      failures.length === 1 ? '' : 's',
+      ' ',
+      formatGeneratedShields(failures, root),
+      '. ',
+      hasTypeScriptErrors
+        ? `TypeScript errors (printed above) can be problematic for shield() generation. Fix your TypeScript errors and try again. See ${docsLink} for more information.`
+        : `See ${docsLink} for how to fix this.`
+    ].join(''),
+    { onlyOnce: true }
+  )
+}
+
+function printSuccesses(root: string) {
+  const successes = generatedShields.filter((s) => !s.failed)
+  const hasFailures = successes.length < generatedShields.length
+  if (successes.length > 0) {
+    console.log(
+      [
+        pc.green('✓'),
+        `${successes.length} shield() generated`,
+        (() => {
+          const onlyOne = successes.length === 1
+          if (!hasFailures) {
+            return `covering ${onlyOne ? 'the telefunction' : 'all telefunctions'}:`
+          } else {
+            return `for the telefunction${onlyOne ? '' : 's'}:`
+          }
+        })(),
+        formatGeneratedShields(successes, root)
+      ].join(' ')
+    )
+  }
+}
+
+function formatGeneratedShields(generatedShields: GeneratedShield[], root: string) {
+  return formatList(
+    generatedShields.map(({ telefunctionName, telefuncFilePath }) => {
+      telefuncFilePath = path.relative(root, telefuncFilePath)
+      return `${pc.bold(telefunctionName + '()')} (${telefuncFilePath})`
+    })
+  )
+}
+
+function formatList(list: string[]): string {
+  return new Intl.ListFormat('en').format(list)
+}
+
+function getShieldName(telefunctionName: string) {
+  return `${telefunctionName}Shield`
+}
+
+let typeToShieldSrc: string | undefined
+function getTypeToShieldSrc() {
+  if (!typeToShieldSrc) {
+    try {
+      // For build `dist/`
+      typeToShieldSrc = fs.readFileSync(`${__dirname}/typeToShield.d.ts`).toString()
+    } catch {
+      // For Vitest
+      typeToShieldSrc = fs.readFileSync(`${__dirname}/typeToShield.ts`).toString()
+    }
+  }
+  assert(typeToShieldSrc)
+  assert(typeToShieldSrc.includes('SimpleType'))
+  return typeToShieldSrc
+}
+
+function findTsConfig(telefuncFilePath: string): string | null {
```

Comment:
This will not work as intended when people use Telefunc with a project that uses [configuration file inheritance](https://www.typescriptlang.org/docs/handbook/project-references.html#overall-structure).

Reply:
Maybe this is a limitation we can address in the documentation, or perhaps `ts-morph` has some tools that can bed used here.

Reply:
I'm probably missing something here, but the way I see it is that it should work because of the following.

From https://www.typescriptlang.org/docs/handbook/tsconfig-json.html:

> The presence of a tsconfig.json file in a directory indicates that the directory is the root of a TypeScript project.

So AFAICT one `tsconfig.json` = one TS project. With the only exception that a TS project may not have any `tsconfig.json` (this PR also handles this case).